commitizen/version_schemes.py

```diff
@@ -0,0 +1,301 @@
+from __future__ import annotations
+
+from itertools import zip_longest
+import re
+import sys
+
+from typing import TYPE_CHECKING, ClassVar, Optional, Type, Tuple, cast
+import warnings
+from commitizen.config.base_config import BaseConfig
+from commitizen.exceptions import VersionSchemeUnknown
+
+import importlib_metadata as metadata
+from packaging.version import Version as _BaseVersion
+from packaging.version import InvalidVersion  # noqa: F401: Rexpose the common exception
+from commitizen.defaults import MAJOR, MINOR, PATCH
+
+if sys.version_info >= (3, 8):
+    from typing import Protocol
+else:
+    from typing_extensions import Protocol
+
+if TYPE_CHECKING:
+    # TypeAlias is Python 3.10+ but backported in typing-extensions
+    if sys.version_info >= (3, 10):
+        from typing import TypeAlias
+    else:
+        from typing_extensions import TypeAlias
+
+    # Self is Python 3.11+ but backported in typing-extensions
+    if sys.version_info < (3, 11):
+        from typing_extensions import Self
+
+
+DEFAULT_VERSION_PARSER = r"v?(?P<version>([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+)?(\w+)?)"
+
+
+class VersionProtocol(Protocol):
```

Comment:
I was thinking the `VersionProvider(ABC)` is an `ABC` instead of `Protocol`, should we stick to ABC for consistency?

Reply:
btw how does `ABC` work with type inference? I'd like to have something that can type hint (this helps during development *if* you use mypy or similar), and that can also fail at runtime if not implemented 🤔 (this for those who don't use any kind of tpye hinting)

Reply:
We need a compromise here:
- `ABC` forces you into inheritance, which is defeating protocol purpose (especially https://github.com/commitizen-tools/commitizen/pull/733#discussion_r1182835871) and it would force `Pep440` and `SemVer` into multiple inheritance
- `Protocol` is the opposite, it forces into nothing because by default it is ignored at runtime

However, there is one possibility which is interesting: we can make the protocol [runtime checkable](https://docs.python.org/3/library/typing.html#typing.runtime_checkable) and raise a warning or fail when calling `get_version_scheme`, something like:

```python
scheme = ep.load()
try:
    isinstance(scheme, VersionProtocol)
except TypeError:
    warnings.warn(f"{name} is not implementing the Version protocol")
return scheme
```

This way we would have best of both world:
- clean type hinting
- freedom of implementation
- runtime error or warning for devs not using MyPy

What do you thing of this solution ?

Reply:
I think it's great. I would also like to choose one between protocol and ABC. Do we need then to keep `versionProvider` as an `ABC`? or can we migrate (at some point in the future) to Protocol with runtime check?

Reply:
OK, it has been added in the latest push (with some tests)

I think there is nothing preventing from switching between `ABC` and `Protocol`. In both cases, this is a design choice, the same as above 👆🏼

Reply:
Totally, I just think we should stick to one for plugins, to reduce cognitive load on those who will implement them.